lib/ProductOpener/Ecoscore.pm

```diff
@@ -472,24 +476,105 @@ sub load_ecoscore_data_packaging() {
-			= $ecoscore_data{packaging_materials}{"en:colored-pet.en:bottle"}{score};
-		$ecoscore_data{packaging_materials}{"en:pet-polyethylene-terephthalate.en:bottle"}
-			= $ecoscore_data{packaging_materials}{"en:colored-pet.en:bottle"};
-		$properties{"packaging_materials"}{"en:pet-polyethylene-terephthalate.en:bottle"}{"ecoscore_score:en"}
-			= $ecoscore_data{packaging_materials}{"en:colored-pet.en:bottle"}{score};
-
-		# Assign transparent rPET bottle score to rPET
-		$ecoscore_data{packaging_materials}{"en:rpet-recycled-polyethylene-terephthalate"}
-			= $ecoscore_data{packaging_materials}{"en:transparent-rpet.en:bottle"};
-		$properties{"packaging_materials"}{"en:rpet-recycled-polyethylene-terephthalate"}{"ecoscore_score:en"}
-			= $ecoscore_data{packaging_materials}{"en:transparent-rpet.en:bottle"}{score};
-
-		$ecoscore_data{packaging_materials}{"en:plastic"} = $ecoscore_data{packaging_materials}{"en:other-plastics"};
-		$properties{"packaging_materials"}{"en:plastic"}{"ecoscore_score:en"}
-			= $ecoscore_data{packaging_materials}{"en:plastic"}{score};
+		# We assign the same score to some target material.shape as a source material.shape
+		# Use English names for source / target shapes and materials
+		# they will be canonicalized with the taxonomies
+		my @assignments = (
+			{
+				target_shape => "bottle",
+				target_material => "opaque pet",
+				source_shape => "bottle",
+				source_material => "colored pet"
+			},
+			{
+				target_shape => "bottle",
+				target_material => "polyethylene terephthalate",
+				source_shape => "bottle",
+				source_material => "colored pet"
+			},
+			# Assign transparent rPET bottle score to rPET
+			{
+				target_shape => "bottle",
+				target_material => "rpet",
+				source_shape => "bottle",
+				source_material => "transparent pet"
+			},
+			{
+				target_material => "plastic",
+				source_material => "other plastics"
+			},
+		);
+
+		foreach my $assignment_ref (@assignments) {
+
+			# We canonicalize the names given in the assignments, as the taxonomies can change over time, including the canonical names
+			my $target_material_id_exists_in_taxonomy;
+			my $target_material = canonicalize_taxonomy_tag(
+				"en", "packaging_materials",
+				$assignment_ref->{target_material},
+				\$target_material_id_exists_in_taxonomy
+			);
+
+			my $source_material_id_exists_in_taxonomy;
+			my $source_material = canonicalize_taxonomy_tag(
+				"en", "packaging_materials",
+				$assignment_ref->{source_material},
+				\$source_material_id_exists_in_taxonomy
+			);
+
+			if (not $target_material_id_exists_in_taxonomy) {
+				die(      "target_material "
+						. $assignment_ref->{target_material}
+						. " does not exist in the packaging_materials taxonomy");
+			}
+			if (not $source_material_id_exists_in_taxonomy) {
+				die(      "source_material "
+						. $assignment_ref->{source_material}
+						. " does not exist in the packaging_materials taxonomy");
+			}
```

Comment:
Although the exists in taxonomy parameter make sense, I don't think you get my point with the refactoring. For me it was more that you are doing 4 times the same thing…

I would have reduced that to:

```perl
  my $target_material = get_canonical_or_die($assignement_ref, "target_material", "packaging_material");
  my $source_material  = get_canonical_or_die($assignement_ref, "source_material", "packaging_material");
  my $target_shape = get_canonical_or_die($assignement_ref, "target_shape", "packaging_shapes");
  my $source_shape = get_canonical_or_die($assignement_ref, "source_shape", "packaging_shapes");
```

Reply:
Sounds good, I added a function like that. I did the other because there are many other places in the code where we canonicalize a tag and then check its existence in the taxonomy (e.g. for all the parsing of ingredients lists etc.)